python/lib/core/dmod/core/dataset.py

```diff
@@ -487,6 +488,252 @@ def unlink_to_dataset(self, dataset: Dataset) -> bool:
         return dataset.manager is not None and dataset.manager.unlink_user(user=self, dataset=dataset)
 
 
+DataItem = TypeVar('DataItem', bound=Union[bytes, RepeatableReader, Path])
+
+
+class AbstractDomainDetector(ABC):
+    """ Abstraction for something that will automatically detect a :class:`DataDomain` for some data. """
+
+    def __init__(self, *args, **kwargs):
```

Comment:
I recommend that we drop this in favor of subtypes defining their own initializer that ideally is the only initializer (no need to call `super().__init__()`). In doing so, subtypes can more freely specify their requirements and loosen the coupling between the super and subclass.

Reply:
Hmm ... I'm not really making use of it, but I had wanted a lazily initialized property for accessing the domain value, and for all the detector types to have this.  My concern was if a detector needed to be used repeatedly (for the same data) that it'd be a performance hit to have to re-detect things over and over.  Though at this point I'm not sure though how often, if ever, that scenario is actually going to come up.  

Is that - complications from introducing a common, inherited property implementation - specifically what you are worried about here, or is there something more?

Reply:
> but I had wanted a lazily initialized property for accessing the domain value, and for all the detector types to have this. My concern was if a detector needed to be used repeatedly (for the same data) that it'd be a performance hit to have to re-detect things over and over.

Yeah, I like that functionality and in some if not all cases that is desirable. My thinking is that subclasses can handle caching / lazy loading in their `detect` implementation. It likely will result in a minor amount of code duplication or at least code that looks repetitive, however I think considering the tradeoffs of having to know the relationship between the parent class and the subclass or a few lines of repetitive code I would rather have as much of the behavior in one place. Another, at least to me, interesting philosophical question. What are your thoughts?